### DISCUSSION TITLE
code-server crashing after update on Raspberry Pi 5

### DISCUSSION THREAD

**Discussion by stepanprokop** [coder/code-server#6690] [General]
Hi everyone!

<img width="538" alt="Screenshot 2024-02-26 at 22 43 46" src="https://github.com/coder/code-server/assets/7387282/c44a881b-ff96-42b3-9642-c806c7f757f0">

I've been using code-server on my Raspberry Pi 5 Model B Rev without any issues until I recently pulled an update. After the update, code-server crashes intermittently, rendering it inaccessible via the web browser. However, it spontaneously becomes available again after about a minute.

**Additional Information:**
Raspberry Pi Model: Raspberry Pi 5 Model B Rev
Operating System: Debian GNU/Linux 12 (bookworm)

**Steps Taken to Troubleshoot:**

1. Checked for existing code-server instances using the command ps aux | grep code-server. Identified two running instances.
2. Attempted to stop the existing code-server processes using the kill command, but the issue persists.
3. Checked for processes using port 8080 using the command sudo netstat -tuln | grep 8080. Found a process listening on port 8080.
4. Checked the version of code-server using the command code-server --version: 4.21.1 45164bf749515cf826e3ccdaa3d45c714dc55e63 with Code 1.86.1.
5. Checked the installed code-server version using the command apt list --installed | grep code-server: code-server/now 4.21.1 arm64 [installed,local].

Feel free to reach out if you need any further information or if there are specific troubleshooting steps I should try.

Thank you!

**Comment by code-asher** [comment]
I would start with looking at code-server's output running with `--log debug`.  

Is it running out of resources?  Memory maybe?  Can you monitor resource usage while reproducing the hang?

Two processes is normal, at minimum you should see a parent code-server process and child process.  But you should see more than two usually.  One for the extension host if the browser is currently connected, another for terminals if any are open, and maybe some others.  If you see only two it could suggest the extension host is crashing then being restarted.  

Just as a reference here is what I have:

```
coder@dev27:~/coder$ ps -eFjH | grep code-server | grep -v grep
coder        925       1     157     157  0 180938 66164 31 21:13 ?        00:00:00   /tmp/code-server/lib/code-server-4.21.1/lib/node /tmp/code-server/lib/code-server-4.21.1 --auth none --port 13337
coder       1099     925     157     157  0 252724 187448 34 21:13 ?       00:00:50     /tmp/code-server/lib/code-server-4.21.1/lib/node /tmp/code-server/lib/code-server-4.21.1/out/node/entry
coder      21964    1099     157     157  0 233527 64104  3 21:40 ?        00:00:03       /tmp/code-server/lib/code-server-4.21.1/lib/node /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/bootstrap-fork --type=ptyHost --logsPath /home/coder/.local/share/code-server/logs/20240226T211825
coder      21975   21964   21975   21975  0  1497  5144  17 21:40 pts/1    00:00:00         /nix/var/nix/profiles/default/bin/bash --init-file /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
coder      24141   21964   24141   24141  0  1497  5072  44 21:42 pts/0    00:00:00         /nix/var/nix/profiles/default/bin/bash --init-file /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
coder      32056   21964   32056   32056  0  1497  5324   6 21:49 pts/2    00:00:00         /nix/var/nix/profiles/default/bin/bash --init-file /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
coder      64503   21964     157     157  0   721   964   2 22:21 ?        00:00:00         /bin/sh -c "/tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/vs/base/node/cpuUsage.sh" 21975
coder      64504   64503     157     157  0  1089  3224  19 22:21 ?        00:00:00           /bin/bash /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/vs/base/node/cpuUsage.sh 21975
coder      64354    1099     157     157  0 179974 57408 44 22:21 ?        00:00:00       /tmp/code-server/lib/code-server-4.21.1/lib/node /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/bootstrap-fork --type=fileWatcher
coder      64382    1099     157     157  0 313184 415984 15 22:21 ?       00:00:03       /tmp/code-server/lib/code-server-4.21.1/lib/node --dns-result-order=ipv4first /tmp/code-server/lib/code-server-4.21.1/lib/vscode/out/bootstrap-fork --type=extensionHost --transformURIs --useHostProxy=false
```

**Comment by stepanprokop** [comment]
Thank you for your guidance and suggestions!

**Start code-server with Debug Logging:**
I initiated code-server with the --log debug option as suggested. Here's the log output during the startup process:
```
code-server --log debug
[2024-02-27T08:56:48.095Z] debug parent:460813 spawned child process 460831
[2024-02-27T08:56:48.472Z] debug child:460831 initiating handshake
[2024-02-27T08:56:48.475Z] debug parent:460813 got message {"message":{"type":"handshake"}}
[2024-02-27T08:56:48.478Z] debug child:460831 got message {"message":{"type":"handshake","args":{"bind-addr":"0.0.0.0:8080","auth":"password","password":"<redacted>","config":"/home/stepanprokop/.config/code-server/config.yaml","log":"debug","user-data-dir":"/home/stepanprokop/.local/share/code-server","extensions-dir":"/home/stepanprokop/.local/share/code-server/extensions","session-socket":"/home/stepanprokop/.local/share/code-server/code-server-ipc.sock","verbose":false,"host":"0.0.0.0","port":8080,"proxy-domain":[],"_":[],"usingEnvPassword":false,"usingEnvHashedPassword":false}}}
[2024-02-27T08:56:48.479Z] info  code-server 4.21.1 45164bf749515cf826e3ccdaa3d45c714dc55e63
[2024-02-27T08:56:48.480Z] info  Using user-data-dir /home/stepanprokop/.local/share/code-server
[2024-02-27T08:56:48.480Z] debug Using extensions-dir /home/stepanprokop/.local/share/code-server/extensions
[2024-02-27T08:56:48.486Z] error listen EADDRINUSE: address already in use 0.0.0.0:8080
[2024-02-27T08:56:48.487Z] debug child:460831 disposing {}
[2024-02-27T08:56:48.491Z] debug parent:460813 inner process 460831 exited unexpectedly
[2024-02-27T08:56:48.492Z] debug parent:460813 disposing {}
```

**Resource Monitoring:**
While reproducing the crash, I closely monitored resource usage on the Raspberry Pi 5 Model B Rev. The system is not experiencing resource depletion; only approximately 1.5% of memory is utilized, and the CPU remains minimally active.

Process Analysis:
```
stepanp+  457693  457677  457693  457693  0 45043 61760   2 09:41 ?        00:00:00     /usr/lib/code-server/lib/node /usr/lib/code-server
stepanp+  457729  457693  457693  457693  1 65927 113648  1 09:41 ?        00:00:17       /usr/lib/code-server/lib/node /usr/lib/code-server/out/node/entry
stepanp+  461999  457729  457693  457693  9 48714 65552   0 10:00 ?        00:00:00         /usr/lib/code-server/lib/node --dns-result-order=ipv4first /usr/lib/code-server/lib/vscode/out/bootstrap-fork --type=extensionHost --transformURIs --useHostProxy=false
```

Thank you for your assistance and guidance throughout this troubleshooting process!

**Reply by code-asher** [reply]
Interesting, the logs say there is already something listening on 8080 so we did not get any information there.  Maybe there was already a code-server process running?  Try killing them all (`pkill -f code-server` or something like that), then run again with `--log debug`, reproduce the issue, and post the results again!
